pkg/agent/flexvolume/controller.go

```diff
@@ -0,0 +1,341 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Some of the code below came from https://github.com/coreos/etcd-operator
+which also has the apache 2.0 license.
+*/
+
+// Package flexvolume to manage Kubernetes storage attach events.
+package flexvolume
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/rook/rook/pkg/agent/flexvolume/crd"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/agent"
+	"github.com/rook/rook/pkg/operator/cluster"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	StorageClassKey       = "storageClass"
+	PoolKey               = "pool"
+	ImageKey              = "image"
+	kubeletDefaultRootDir = "/var/lib/kubelet"
+)
+
+var driverLogger = capnslog.NewPackageLogger("github.com/rook/rook", "rook-flexdriver")
+
+// FlexvolumeController handles all events from the Flexvolume driver
+type FlexvolumeController struct {
+	clientset                  kubernetes.Interface
+	volumeManager              VolumeManager
+	volumeAttachmentController *crd.VolumeattachmentController
+}
+
+func newFlexvolumeController(context *clusterd.Context, volumeAttachmentCRDClient rest.Interface, manager VolumeManager) *FlexvolumeController {
+	return &FlexvolumeController{
+		clientset:                  context.Clientset,
+		volumeManager:              manager,
+		volumeAttachmentController: crd.New(volumeAttachmentCRDClient),
+	}
+}
+
+// Attach attaches rook volume to the node
+func (c *FlexvolumeController) Attach(attachOpts AttachOptions, devicePath *string) error {
+
+	namespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	node := os.Getenv(k8sutil.NodeNameEnvVar)
+
+	// Name of CRD is the PV name. This is done so that the CRD can be use for fencing
+	crdName := attachOpts.VolumeName
+
+	// Check if this volume has been attached
+	volumeattachObj, err := c.volumeAttachmentController.Get(namespace, crdName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get volume CRD %s. %+v", crdName, err)
+		}
+		// No volumeattach CRD for this volume found. Create one
+		volumeattachObj = crd.NewVolumeattachment(crdName, namespace, node, attachOpts.PodNamespace, attachOpts.Pod, attachOpts.MountDir, strings.ToLower(attachOpts.RW) == "ro")
```

Comment:
consider some line breaks one these longer lines, i'm having to scroll to read them